src/hover_provider/modelHoverProvider.ts

```diff
@@ -0,0 +1,161 @@
+import {
+  CancellationToken,
+  HoverProvider,
+  Disposable,
+  Position,
+  ProviderResult,
+  Range,
+  TextDocument,
+  Uri,
+  Hover,
+  MarkdownString,
+} from "vscode";
+import { NodeMetaMap } from "../domain";
+import { DBTProjectContainer } from "../manifest/dbtProjectContainer";
+import { ManifestCacheChangedEvent } from "../manifest/event/manifestCacheChangedEvent";
+import { provideSingleton } from "../utils";
+import { TelemetryService } from "../telemetry";
+
+@provideSingleton(ModelHoverProvider)
+export class ModelHoverProvider implements HoverProvider, Disposable {
+  private modelToLocationMap: Map<string, NodeMetaMap> = new Map();
+  private static readonly IS_REF = /(ref)\([^)]*\)/;
+  private static readonly GET_DBT_MODEL = /(?!'|")([^(?!'|")]*)(?='|")/gi;
+  private disposables: Disposable[] = [];
+
+  constructor(
+    private dbtProjectContainer: DBTProjectContainer,
+    private telemetry: TelemetryService,
+  ) {
+    this.disposables.push(
+      dbtProjectContainer.onManifestChanged((event) =>
+        this.onManifestCacheChanged(event),
+      ),
+    );
+  }
+
+  dispose() {
+    while (this.disposables.length) {
+      const x = this.disposables.pop();
+      if (x) {
+        x.dispose();
+      }
+    }
+  }
+
+  provideHover(
+    document: TextDocument,
+    position: Position,
+    token: CancellationToken,
+  ): ProviderResult<Hover> {
+    return new Promise((resolve) => {
+      const hover = document.getText(document.getWordRangeAtPosition(position));
+      const word = document.getText(
+        document.getWordRangeAtPosition(position, ModelHoverProvider.IS_REF),
+      );
+      const project = this.dbtProjectContainer.findDBTProject(document.uri);
+      if (!project) {
+        console.error(
+          "Could not load hover provider, project not found in container for " +
+            document.uri.fsPath,
+        );
+        return;
+      }
+      if (word !== undefined && hover !== "ref") {
+        const dbtModel = word.match(ModelHoverProvider.GET_DBT_MODEL);
```

Comment:
Is it possible to extract this logic to a shared utils class?

This would make it easier to evolve this code and fix issues in both the definition provider and the hover provider.

Reply:
Ya - there is definitely code duplication here. I was hoping to limit this to a v1 sort of feature but I can pull some of this out into shared utility.

Reply:
It was just a question, let's do this as a follow up.